File: aplikacijaFlutter/client/lib/trade_execution.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;

class TradeRequest {
    int idRaz = 1;
    int idUser = 1;
    List<int> offered = [];
    List<int> wanted = [];
    int acceptDuplicates = -1;

    TradeRequest(this.idRaz, this.idUser, this.acceptDuplicates){}

    @override
    String toString(){
      return idRaz.toString() + " " + idUser.toString() + " " + offered.toString() + " " + wanted.toString();
    }


}

class TradeToExecute {
  int firstId = -1;
  int secondId = -1;
  List<int> firstToSecond = [];
  List<int> secondToFirst = [];
  int firstUser = -1;
  int secondUser = -1;
  int firstPlaceInTrade = -1;
  int secondPlaceInTrade = -1;

  TradeToExecute(this.firstId, this.secondId, this.firstToSecond, this.secondToFirst, this.firstUser, this.secondUser, this.firstPlaceInTrade, this.secondPlaceInTrade){}

  @override
  String toString() {
    return "First Id: " + firstId.toString() + "\n Second Id: " + secondId.toString() + "\n First to Second: " + firstToSecond.toString() +
    "\n Second to first:" + secondToFirst.toString() + "\n First user: " + firstUser.toString() + "\n Second user: " + secondUser.toString();
  }
}

List<int> removeFirstOccurance(List<int> firstList, List<int> secondList){
   Map<int, int> secondMap = {};

  for (int element in secondList) {
    secondMap[element] = (secondMap[element] ?? 0) + 1;
  }

  firstList.removeWhere((element) {
    if (secondMap.containsKey(element) && secondMap[element]! > 0) {
      secondMap[element] = secondMap[element]! - 1;
      return true;
    }
    return false;
  });

  return firstList; 
}

class TradeExecutionTab extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: ElevatedButton(
        onPressed: () async {
          List<TradeRequest> allRequests = [];
          
          final response = await http.get(
          Uri.parse('http://localhost:3001/stickers/dohvatiRazmene'),
          );
          final Map<dynamic, dynamic> responseData = jsonDecode(response.body);
          final List<dynamic> data = responseData['razmene'] ?? [];
          for(int i = 0; i < data.length; i++){
              allRequests.add(TradeRequest(data[i]['idRaz'], data[i]['idUser'], data[i]['acceptDuplicates']));
          }

          final response2 = await http.get(
          Uri.parse('http://localhost:3001/stickers/dohvatiPonude'),
          );

          final Map<dynamic, dynamic> responseData2 = jsonDecode(response2.body);
          final List<dynamic> data2 = responseData2['ponude'] ?? [];


          for(int i = 0; i < data2.length; i++){
            TradeRequest requestToUpdate = allRequests.firstWhere((request) => request.idRaz == data2[i]['idRaz']);
            requestToUpdate.offered.add(data2[i]['idIgr']);
          }

          final response3 = await http.get(
          Uri.parse('http://localhost:3001/stickers/dohvatiZelje'),
          );

          final Map<dynamic, dynamic> responseData3 = jsonDecode(response3.body);
          final List<dynamic> data3 = responseData3['zelje'] ?? [];


          for(int i = 0; i < data3.length; i++){
            TradeRequest requestToUpdate = allRequests.firstWhere((request) => request.idRaz == data3[i]['idRaz']);
            requestToUpdate.wanted.add(data3[i]['idIgr']);
          }

          

          /*

          for(int i = 0; i < 10; i++){
            allRequests.add(TradeRequest(i, i));
          }

          Random random = Random();
          
          for(int i = 0; i < allRequests.length; i++){
            List<int> numbers = [];
            for(int i = 0; i < 10; i++){
              numbers.add(random.nextInt(100) + 1);
            }
            allRequests[i].offered = numbers;
          }

          for(int i = 0; i < allRequests.length; i++){
            List<int> numbers = [];
            for(int i = 0; i < 10; i++){
              numbers.add(random.nextInt(100) + 1);
            }
            allRequests[i].wanted = numbers;
          }

          for(int i = 0; i < allRequests.length; i++){
            Set<int> numbers = Set<int>();
            while(numbers.length < 10){
              numbers.add(random.nextInt(100) + 1);
            }
            allRequests[i].wanted = numbers.toList();
          }

          for(int i = 0; i < allRequests.length; i++){
            allRequests[i].offered.removeWhere((element) => allRequests[i].wanted.contains(element));
          }

          */

          bool isComplete = false;

          List<TradeToExecute> trades = [];

          int beginTradesToMake = 0;

          

          while(!isComplete){
            isComplete = true;
            Set<int> checked = Set<int>();
            for(int i = 0; i < allRequests.length - 1; i++){
              if(checked.contains(allRequests[i].idRaz)) continue;
              checked.add(allRequests[i].idRaz);
              int maxId = -1;
              int maxValue = 0;
              List<int> solutionList1 = [];
              List<int> solutionList2 = [];
              for(int j = i + 1; j < allRequests.length; j++){
                if(checked.contains(allRequests[j].idRaz)) continue;
                Set<int> firstElementOffered = allRequests[i].offered.toSet();
                Set<int> firstElementWanted = allRequests[i].wanted.toSet();
                Set<int> secondElementOffered = allRequests[j].offered.toSet();
                Set<int> secondElementWanted = allRequests[j].wanted.toSet();

                firstElementOffered.removeWhere((element) => !secondElementWanted.contains(element));
                secondElementOffered.removeWhere((element) => !firstElementWanted.contains(element));

                int max = firstElementOffered.length < secondElementOffered.length ? firstElementOffered.length : secondElementOffered.length;

                if(max > maxValue){
                  if(isComplete) isComplete = false;
                  maxValue = max;
                  maxId = j;
                  solutionList1 = firstElementOffered.toList();
                  solutionList1 = solutionList1.sublist(0, max);
                  solutionList2 = secondElementOffered.toList();
                  solutionList2 = solutionList2.sublist(0, max);
                }
              }

              if(maxId != -1){
                checked.add(allRequests[maxId].idRaz);
                trades.add(TradeToExecute(allRequests[i].idRaz, allRequests[maxId].idRaz, solutionList1, solutionList2, allRequests[i].idUser, allRequests[maxId].idUser, i, maxId));
              }

              
            }

            

            for(int i = beginTradesToMake; i < trades.length; i++){
                allRequests[trades[i].firstPlaceInTrade].offered = removeFirstOccurance(allRequests[trades[i].firstPlaceInTrade].offered, trades[i].firstToSecond);
                allRequests[trades[i].secondPlaceInTrade].wanted = removeFirstOccurance(allRequests[trades[i].secondPlaceInTrade].wanted, trades[i].firstToSecond);
                allRequests[trades[i].firstPlaceInTrade].wanted = removeFirstOccurance(allRequests[trades[i].firstPlaceInTrade].wanted, trades[i].secondToFirst);
                allRequests[trades[i].secondPlaceInTrade].offered = removeFirstOccurance(allRequests[trades[i].secondPlaceInTrade].offered, trades[i].secondToFirst);
              }

              beginTradesToMake = trades.length;
            
          }

          for(int i = 0; i < allRequests.length; i++){
            if(allRequests[i].offered.isEmpty || allRequests[i].wanted.isEmpty || allRequests[i].acceptDuplicates == 0) continue;
            for(int j = 0; j < allRequests.length; j++){
              if(i == j) continue;
              if(allRequests[j].offered.isEmpty || allRequests[j].acceptDuplicates == 0) continue;
              Set<int> firstElementWanted = allRequests[i].wanted.toSet();
              Set<int> secondElementOffered = allRequests[j].offered.toSet();
              secondElementOffered.removeWhere((element) => !firstElementWanted.contains(element));
              if(secondElementOffered.isEmpty) continue;
              List<int> solutionList1 = [];
              List<int> solutionList2 = [];
              int max = secondElementOffered.length < allRequests[i].offered.length ? secondElementOffered.length : allRequests[i].offered.length;
              solutionList1 = allRequests[i].offered;
              solutionList1 = solutionList1.sublist(0, max);
              solutionList2 = secondElementOffered.toList();
              solutionList2 = solutionList2.sublist(0, max);
              TradeToExecute newTrade = TradeToExecute(allRequests[i].idRaz, allRequests[j].idRaz, solutionList1, solutionList2, allRequests[i].idUser, allRequests[j].idUser, i, j);
              print(newTrade);
              trades.add(newTrade);
              print(trades[trades.length - 1]);
              allRequests[i].offered = removeFirstOccurance(allRequests[i].offered, solutionList1);
              allRequests[i].wanted = removeFirstOccurance(allRequests[i].wanted, solutionList2);
              allRequests[j].offered = removeFirstOccurance(allRequests[j].offered, solutionList2);
              for(int k = 0; k < solutionList2.length; k++){ 
                allRequests[j].offered.add(solutionList2[k]);
              }
              


            }
          }

        
          final headers = {
            'Content-Type': 'application/json; charset=UTF-8',
          };

          for(int i = 0; i < trades.length; i++){
            print(i);
            print(trades[i].toString());
          }

          
          

          
          
          for(int i = 0; i < trades.length; i++){
            for(int j = 0; j < trades[i].firstToSecond.length; j++){
              final response1 = await http.post(Uri.parse("http://localhost:3001/stickers/dohvatiSlicicuZaRazmenu"),
              headers: headers,
              body: jsonEncode({
                'idIgr': trades[i].firstToSecond[j],
                'idRaz': trades[i].firstId,
              }));
              final response1data = jsonDecode(response1.body);
              final List<dynamic> data1 = response1data['slicica'] ?? [];
              int idNez1 = data1[0]['idNez'];

        

              
              final response2 = await http.post(Uri.parse("http://localhost:3001/stickers/promeniVlasnikaSlicice"),
              headers: headers,
              body: jsonEncode({
                'idNez': idNez1,
                'idUser': trades[i].secondUser,
              }));

              final response3 = await http.post(Uri.parse("http://localhost:3001/stickers/obrisiSlicicuIzZamene"),
              headers: headers,
              body: jsonEncode({
                'idNez': idNez1,
              }));

              final response4 = await http.post(Uri.parse("http://localhost:3001/stickers/dohvatiSlicicuZaRazmenu"),
              headers: headers,
              body: jsonEncode({
                'idIgr': trades[i].secondToFirst[j],
                'idRaz': trades[i].secondId,
              }));
              final response4data = jsonDecode(response4.body);
              final List<dynamic> data4 = response4data['slicica'] ?? [];
              int idNez2 = data4[0]['idNez'];

              final response5 = await http.post(Uri.parse("http://localhost:3001/stickers/promeniVlasnikaSlicice"),
              headers: headers,
              body: jsonEncode({
                'idNez': idNez2,
                'idUser': trades[i].firstUser,
              }));

              final response6 = await http.post(Uri.parse("http://localhost:3001/stickers/obrisiSlicicuIzZamene"),
              headers: headers,
              body: jsonEncode({
                'idNez': idNez2,
              }));

              

              

            }
          }

          final response7 = await http.get(Uri.parse("http://localhost:3001/stickers/obrisiRazmene"));
          
          
        },
        child: Text('Execute Trade'),
      ),
    );
  }
} 

File: aplikacijaReact/client/src/pages/TradeExecution.js
import axios from 'axios';
import React, { useEffect, useState } from 'react';

class TradeRequest {
    constructor(idRaz, idUser, acceptDuplicates) {
        this.idRaz = idRaz;
        this.idUser = idUser;
        this.wanted = [];
        this.offered = [];
        this.acceptDuplicates = acceptDuplicates;
    }
}

const TradeExecution = () => {
    const [requests, setRequests] = useState([]);
    const [start, setStart] = useState(0);

    function generateRandomList(mustBeUnique) {
        let list = [];
        while (list.length < 20) {
            let randomNumber = Math.floor(Math.random() * 100) + 1;
            if (!list.includes(randomNumber) || !mustBeUnique) {
                list.push(randomNumber);
            }
        }
        return list;
    }

    function removeCommonItems(arr1, arr2) {
        let commonItems = arr1.filter(item => arr2.includes(item));
        commonItems.forEach(item => {
            arr1.splice(arr1.indexOf(item), 1);
        });
        return arr1;
    }

    function removeFirstOccurrences(list1, list2) {
        const map = {};
        for (const item of list2) {
            map[item] = true;
        }
        return list1.filter(item => {
            if (map[item]) {
                delete map[item];
                return false;
            }
            return true;
        });
    }

    useEffect(() => {
        const fetchData = async () => {

            
            
            try {
                const response2 = await axios.get("http://localhost:3001/stickers/dohvatiPonude");
                const response3 = await axios.get("http://localhost:3001/stickers/dohvatiZelje");
                const response = await axios.get("http://localhost:3001/stickers/dohvatiRazmene");

                let tempList = response.data.razmene.map(item => new TradeRequest(item.idRaz, item.idUser, item.acceptDuplicates));
                setRequests(tempList);



                // Update TradeRequest instances with idIgr from response2
                setRequests(tempList => {
                    const updatedRequests = tempList.map(request => {
                        const matchingPonude = response2.data.ponude.filter(ponuda => ponuda.idRaz === request.idRaz);
                        if (matchingPonude.length > 0 && request.offered.length == 0) {
                            request.offered.push(...matchingPonude.map(ponuda => ponuda.idIgr));
                        }
                        return request;
                    });
                
                    tempList = updatedRequests; // Update tempList
                    return updatedRequests;
                });

                setRequests(tempList => {
                    const updatedRequests = tempList.map(request => {
                        const matchingZelje = response3.data.zelje.filter(zelja => zelja.idRaz === request.idRaz);
                        if (matchingZelje.length > 0 && request.wanted.length == 0) {
                            request.wanted.push(...matchingZelje.map(zelja => zelja.idIgr));
                        }
                        return request;
                    });
                
                    tempList = updatedRequests; // Update tempList
                    return updatedRequests;
                });


                setStart(1);


            } catch (error) {
                console.error("Error fetching data:", error);
            }
            

            // using test data
/*
            let tempList = []


            for(let i = 1; i <= 10; i++){
                tempList.push(new TradeRequest(i, i));
            }
            
            
            for(let i = 0; i < 10; i++){

                tempList[i].offered = generateRandomList(false);
                tempList[i].wanted = generateRandomList(true);
                tempList[i].offered = removeCommonItems(tempList[i].offered, tempList[i].wanted);

            }

            console.log(JSON.stringify(tempList))

*/

            
          

            

            
        };

        fetchData(); // Call the fetchData function to initiate the data fetching
    }, []); // The empty dependency array ensures the effect runs only once

    useEffect(() => {
        const execute = async () => {   

            if(start == 0) return;

            let completedList = [];

            let tradesToMake = [];

            let isComplete = false;

            let beginTradesToMake = 0;

        //    for(let i = 0; i < requests.length; i++) console.log(JSON.stringify(requests[i])); 

            while(!isComplete){


                isComplete = true;

                for(let i = 0; i < requests.length - 1; i++){
                    if (completedList.includes(requests[i].idRaz)) continue;
                    completedList.push(requests[i].idRaz);
                    let maxId = -1;
                    let maxValue = 0;
                    let solutionList1 = []
                    let solutionList2 = []
                    for(let j = i + 1; j < requests.length; j++){
                        if(completedList.includes(requests[j].idRaz)) continue;
                        const firstElementOffered = new Set(requests[i].offered);
                        const firstElementWanted = new Set(requests[i].wanted);
                        const secondElementOffered = new Set(requests[j].offered);
                        const secondElementWanted = new Set(requests[j].wanted);

                        
                        const common1 = Array.from(firstElementOffered).filter(element => secondElementWanted.has(element)); // Corrected
                        const common2 = Array.from(secondElementOffered).filter(element => firstElementWanted.has(element)); // Corrected
                        let max = Math.min(common1.length, common2.length);
    
                        if(max > maxValue) {                            
                            if(isComplete) isComplete = false;
                            maxValue = max;
                            maxId = j;
                            solutionList1 = common1.slice(0, max);
                            solutionList2 = common2.slice(0, max);
                            
                        }
                    }
                    if (maxId != -1){
                        completedList.push(requests[maxId].idRaz);
                        let newSolution = [requests[i].idRaz, requests[maxId].idRaz, solutionList1, solutionList2, requests[i].idUser, requests[maxId].idUser, i, maxId]
                        tradesToMake.push(newSolution);
                        
                    }
                }
    
                for(let i = beginTradesToMake; i < tradesToMake.length; i++){
                    requests[tradesToMake[i][6]].offered = removeFirstOccurrences(requests[tradesToMake[i][6]].offered, tradesToMake[i][2]);
                    requests[tradesToMake[i][6]].wanted = removeFirstOccurrences(requests[tradesToMake[i][6]].wanted, tradesToMake[i][3]);
                    requests[tradesToMake[i][7]].offered = removeFirstOccurrences(requests[tradesToMake[i][7]].offered, tradesToMake[i][3]);
                    requests[tradesToMake[i][7]].wanted = removeFirstOccurrences(requests[tradesToMake[i][7]].wanted, tradesToMake[i][2]);

                }

                completedList = [];

                beginTradesToMake = tradesToMake.length;

            }
            /*
            for(let i = 0; i < tradesToMake.length; i++) console.log(JSON.stringify(tradesToMake[i]))
            for(let i = 0; i < requests.length; i++) console.log(JSON.stringify(requests[i])); 
            */

            

                
            for(let i = 0; i < tradesToMake.length; i++){
                for(let j = 0; j < tradesToMake[i][2].length; j++){

                    console.log(j);

                    
                //    idNez = getIdNez(tradesToMake[i][2][j], tradesToMane[i][0]); // select idNez from ponuda_razmena where idIgr = tradesToMake[i][2][j] and idRaz = tradesToMake[i][0]

                    const response1 = await axios.post("http://localhost:3001/stickers/dohvatiSlicicuZaRazmenu", {idIgr: tradesToMake[i][2][j], idRaz: tradesToMake[i][0]});
                    
                    let idNez1 = response1.data.slicica[0].idNez
                    console.log(idNez1)
                    

                //    changeOwner(idNez, tradesToMake[i][5]); // update nezalepljena set idUser = tradesToMake[i][5] where nezalepljena.idNez = idNez

                    const response2 = await axios.post("http://localhost:3001/stickers/promeniVlasnikaSlicice", {idUser: tradesToMake[i][5], idNez: idNez1});
                    console.log(response2.data)

                //    deleteFromPonuda(idNez) //delete from ponuda_razmena where ponuda_razmena.idNez = idNez
                    
                    const response3 = await axios.post("http://localhost:3001/stickers/obrisiSlicicuIzZamene", {idNez: idNez1});
                    console.log(response3.data)

                //    idNez = getIdNez(tradesToMake[i][3][j], tradesToMake[i][1]); // select idNez from ponuda_razmena where idIgr = tradesToMake[i][3][j] and idRaz = tradesToMake[i][1]

                    const response4 = await axios.post("http://localhost:3001/stickers/dohvatiSlicicuZaRazmenu", {idIgr: tradesToMake[i][3][j], idRaz: tradesToMake[i][1]});
                    
                    let idNez2 = response4.data.slicica[0].idNez
                    console.log(idNez2)

                //    changeOwner(idNez, tradesToMake[i][4]); // update nezalepljena set idUser = tradesToMake[i][4] where nezalepljena.idNez = idNez

                    const response5 = await axios.post("http://localhost:3001/stickers/promeniVlasnikaSlicice", {idUser: tradesToMake[i][4], idNez: idNez2});
                    console.log(response5.data);

                //    deleteFromPonuda(idNez) //delete from ponuda_razmena where ponuda_razmena.idNez = idNez

                    const response6 = await axios.post("http://localhost:3001/stickers/obrisiSlicicuIzZamene", {idNez: idNez2});
                    console.log(response6.data)


                }
            }

            
                
        //    deleteAllTradeRequests();

        for(let i = 0; i < requests.length; i++){
            if(requests[i].offered.length == 0 || requests[i].wanted.length == 0 || requests[i].acceptDuplicates == 0) continue;
            for(let j = 0; j < requests.length; j++){
                if(i == j) continue;
                if(requests[j].offered.length == 0 || requests[j].acceptDuplicates == 0) continue;
                const firstElementWanted = new Set(requests[i].wanted);
                const secondElementOffered = new Set(requests[j].offered);
                const common = Array.from(secondElementOffered).filter(element => firstElementWanted.has(element));
                console.log(common);
                if(common.length == 0) continue;
                let solutionList1 = [];
                let solutionList2 = [];
                let max = Math.min(common.length, requests[i].offered.length);
                solutionList2 = common.slice(0, max);
                solutionList1 = requests[i].offered.slice(0, max);
                console.log(solutionList1);
                console.log(solutionList2);
                let newSolution = [requests[i].idRaz, requests[j].idRaz, solutionList1, solutionList2, requests[i].idUser, requests[j].idUser, i, j];
                requests[i].offered = removeFirstOccurrences(requests[i].offered, solutionList1);
                requests[i].wanted = removeFirstOccurrences(requests[i].wanted, solutionList2);
                requests[j].offered = removeFirstOccurrences(requests[j].offered, solutionList2);
                for(let k = 0; k < solutionList2.length; k++) requests[j].offered.push(solutionList2[k]);
                tradesToMake.push(newSolution);


                

    


            }
        }

        for(let i = beginTradesToMake; i < tradesToMake.length; i++){
            for(let j = 0; j < tradesToMake[i][2].length; j++){

                console.log(j);

                
            //    idNez = getIdNez(tradesToMake[i][2][j], tradesToMane[i][0]); // select idNez from ponuda_razmena where idIgr = tradesToMake[i][2][j] and idRaz = tradesToMake[i][0]

                const response1 = await axios.post("http://localhost:3001/stickers/dohvatiSlicicuZaRazmenu", {idIgr: tradesToMake[i][2][j], idRaz: tradesToMake[i][0]});
                
                let idNez1 = response1.data.slicica[0].idNez
                console.log(idNez1)
                

            //    changeOwner(idNez, tradesToMake[i][5]); // update nezalepljena set idUser = tradesToMake[i][5] where nezalepljena.idNez = idNez

                const response2 = await axios.post("http://localhost:3001/stickers/promeniVlasnikaSlicice", {idUser: tradesToMake[i][5], idNez: idNez1});
                console.log(response2.data)

            //    deleteFromPonuda(idNez) //delete from ponuda_razmena where ponuda_razmena.idNez = idNez
                
                const response3 = await axios.post("http://localhost:3001/stickers/obrisiSlicicuIzZamene", {idNez: idNez1});
                console.log(response3.data)

            //    idNez = getIdNez(tradesToMake[i][3][j], tradesToMake[i][1]); // select idNez from ponuda_razmena where idIgr = tradesToMake[i][3][j] and idRaz = tradesToMake[i][1]

                const response4 = await axios.post("http://localhost:3001/stickers/dohvatiSlicicuZaRazmenu", {idIgr: tradesToMake[i][3][j], idRaz: tradesToMake[i][1]});
                
                let idNez2 = response4.data.slicica[0].idNez
                console.log(idNez2)

            //    changeOwner(idNez, tradesToMake[i][4]); // update nezalepljena set idUser = tradesToMake[i][4] where nezalepljena.idNez = idNez

                const response5 = await axios.post("http://localhost:3001/stickers/promeniVlasnikaSlicice", {idUser: tradesToMake[i][4], idNez: idNez2});
                console.log(response5.data);

            //    deleteFromPonuda(idNez) //delete from ponuda_razmena where ponuda_razmena.idNez = idNez

                const response6 = await axios.post("http://localhost:3001/stickers/obrisiSlicicuIzZamene", {idNez: idNez2});
                console.log(response6.data)


            }
        }

            

            
            const response = await axios.get("http://localhost:3001/stickers/obrisiRazmene");
            console.log("DONE")
            

        };

        execute();

    }, [start]);

    return (
        <div className='content'>
            
        </div>
    );
}

export default TradeExecution;
